homeassistant/components/ring/__init__.py

```diff
@@ -146,98 +156,141 @@ def token_updater(token):
+        self._unsub_interval = None
+
+    @callback
+    def add_listener(self, update_callback):
+        """Listen for data updates."""
+        # This is the first listener, set up interval.
+        if not self.listeners:
+            self._unsub_interval = async_track_time_interval(
+                self.hass, self.async_refresh_all, self.update_interval
+            )
+
+        self.listeners.append(update_callback)
+
+    @callback
+    def remove_listener(self, update_callback):
+        """Remove data update."""
+        self.listeners.remove(update_callback)
+
+        if not self.listeners:
+            self._unsub_interval()
+            self._unsub_interval = None
+
+    async def async_refresh_all(self, _now: Optional[int] = None) -> None:
+        """Time to update."""
+        if not self.listeners:
+            return
+
+        try:
+            await self.hass.async_add_executor_job(
+                getattr(self.ring, self.update_method)
+            )
+        except AccessDeniedError:
+            _LOGGER.error("Ring access token is no longer valid. Set up Ring again")
+            await self.hass.config_entries.async_unload(self.config_entry_id)
+            return
+
+        for update_callback in self.listeners:
+            update_callback()
+
+
+class DeviceDataUpdater:
+    """Data storage for device data."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        config_entry_id: str,
+        ring: Ring,
+        update_method: str,
+        update_interval: timedelta,
+    ):
+        """Initialize device data updater."""
+        self.hass = hass
+        self.config_entry_id = config_entry_id
+        self.ring = ring
+        self.update_method = update_method
+        self.update_interval = update_interval
         self.devices = {}
         self._unsub_interval = None
 
-    async def track_device(self, config_entry_id, device):
+    async def track_device(self, device, update_callback):
         """Track a device."""
         if not self.devices:
             self._unsub_interval = async_track_time_interval(
-                self.hass, self.refresh_all, SCAN_INTERVAL
+                self.hass, self.refresh_all, self.update_interval
             )
 
-        key = (config_entry_id, device.device_id)
-
-        if key not in self.devices:
-            self.devices[key] = {
+        if device.device_id not in self.devices:
+            self.devices[device.device_id] = {
                 "device": device,
-                "count": 1,
+                "update_callbacks": [update_callback],
+                "data": None,
             }
+            # Store task so that other concurrent requests can wait for us to finish and
+            # data be available.
+            self.devices[device.device_id]["task"] = asyncio.current_task()
+            self.devices[device.device_id][
+                "data"
+            ] = await self.hass.async_add_executor_job(self.update_method, device)
```

Comment:
Does tracking a device also need to update the data? Couldn't we just let the consumer get the data from the cache?

Why do we need to let the platforms track devices? We already know what devices there are. We could let the component control the device interface completely and just have platforms get data as needed from the cache.

Reply:
If an entity is disabled, we wouldn't want to fetch that data.